src/Vault.sol

```diff
@@ -658,92 +648,113 @@ contract Vault {
-        vault._baseCurrency = uint8(newBaseCurrency); //Change this to where ever it is going to be actually set
+        require(getUsedMargin() == 0, "VL: Can't change baseCurrency when Used Margin > 0");
+        require(IMainRegistry(registryAddress).isBaseCurrency(_baseCurrency), "VL: baseCurrency not found");
+        vault.baseCurrency = _baseCurrency; //Change this to where ever it is going to be actually set
     }
 
-    // https://twitter.com/0x_beans/status/1502420621250105346
+    /*///////////////////////////////////////////////////////////////
+                    MARGIN ACCOUNT SETTINGS
+    ///////////////////////////////////////////////////////////////*/
+
+    bool isTrustedProtocolSet;
+    address public trustedProtocol;
+
     /** 
-    @notice Returns the sum of all uints in an array.
-    @param _data An uint256 array.
-    @return sum The combined sum of uints in the array.
-  */
-    function sumElementsOfList(uint256[] memory _data)
-        public
-        payable
-        returns (uint256 sum)
-    {
-        //cache
-        uint256 len = _data.length;
+    @notice Initiates a margin account on the vault for one trusted application..
+    @param protocol The contract address of the trusted application.
+    @dev The open position is fetched at a contract of the application -> only allow trusted audited protocols!!!
+    @dev Currently only one trusted protocol can be set.
+    */
+    function openTrustedMarginAccount(address protocol) onlyOwner public {
+        require(!isTrustedProtocolSet, "V_OMA: ALREADY SET");
+        //ToDo: Check in Factory/Mainregistry if protocol is indeed trusted?
 
-        for (uint256 i = 0; i < len; ) {
-            // optimizooooor
-            assembly {
-                sum := add(sum, mload(add(add(_data, 0x20), mul(i, 0x20))))
-            }
+        (bool success, address baseCurrency, address liquidator_) = ITrustedProtocol(protocol).openMarginAccount();
+        require(success, "V_OMA: OPENING ACCOUNT REVERTED");
 
-            unchecked {
-                ++i;
-            }
-        }
+        liquidator = liquidator_;
+        trustedProtocol = protocol;
+        if (vault.baseCurrency != baseCurrency) _setBaseCurrency(baseCurrency);
+        IERC20(baseCurrency).approve(protocol, type(uint256).max);
+        isTrustedProtocolSet = true;
+        allowed[protocol] = true;
+    }
+
+    /** 
+    @notice Closes the margin account on the vault of the trusted application..
+    @dev The open position is fetched at a contract of the application -> only allow trusted audited protocols!!!
+    @dev Currently only one trusted protocol can be set.
+    */
+    function closeTrustedMarginAccount() public onlyOwner {
+        require(isTrustedProtocolSet, "V_CMA: NOT SET");
+
+        require(ITrustedProtocol(trustedProtocol).getOpenPosition(address(this)) == 0, "NON-ZERO OPEN POSITION");
+
+        isTrustedProtocolSet = false;
```

Comment:
We should pack this in a struct

Reply:
I propose we do the packing of variables and other gas optimisations after complete refactor (there are some pending PRs after this one)